### PR TITLE
IGNITE-23487 Support cancellation tokens in IgniteCompute

### PR DESCRIPTION
**Pull request by zstan**
Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/compute/src/main/java/org/apache/ignite/internal/compute/ComputeComponentImpl.java [hunk 1]
```diff
@@ -145,6 +150,14 @@ public <I, R> JobExecution<R> executeLocally(
             inFlightFutures.registerFuture(future);
 
             JobExecution<R> result = new DelegatingJobExecution<>(future);
+
+            if (cancellationToken != null) {
+                CancelHandleHelper.addCancelAction(cancellationToken, () -> classLoaderFut
+                        .cancel(true), classLoaderFut);
```
[the remark this answers is not in the document]
Comment (zstan) [reply modules/compute/src/main/java/org/apache/ignite/internal/compute/ComputeComponentImpl.java:156]: don\`t understand what future you are talking about ?

File: modules/compute/src/main/java/org/apache/ignite/internal/compute/ComputeJobFailover.java [hunk 2]
```diff
@@ -84,6 +86,9 @@ class ComputeJobFailover<R> {
      */
     private final RemoteExecutionContext<?, R> jobContext;
 
+    /** Cancellation token. */
+    @Nullable private final CancellationToken cancellationToken;
```
[the remark this answers is not in the document]
Comment (zstan) [reply modules/compute/src/main/java/org/apache/ignite/internal/compute/ComputeJobFailover.java:90]: i suppose - no, cause ExecutionOptions is passed into execution nodes and if token was passed there - distributed nodes are register redundant token.

File: modules/compute/src/test/java/org/apache/ignite/internal/compute/IgniteComputeImplTest.java [hunk 3]
```diff
@@ -243,19 +268,24 @@ private void respondWhenAskForPrimaryReplica() {
...
 
+    private void respondWhenExecutingSimpleJobLocally(ExecutionOptions executionOptions, CancellationToken cancellationToken) {
+        when(computeComponent.executeLocally(eq(executionOptions), eq(testDeploymentUnits), eq(JOB_CLASS_NAME),
+                eq(cancellationToken), eq("a"))).thenReturn(completedExecution("jobResponse"));
+    }
+
     private void respondWhenExecutingSimpleJobRemotely(ExecutionOptions options) {
         when(computeComponent.executeRemotelyWithFailover(
-                eq(remoteNode), any(), eq(testDeploymentUnits), eq(JOB_CLASS_NAME), eq(options), eq("a")
+                eq(remoteNode), any(), eq(testDeploymentUnits), eq(JOB_CLASS_NAME), eq(options), eq(null), eq("a")
```
[the remark this answers is not in the document]
Comment (zstan) [reply modules/compute/src/test/java/org/apache/ignite/internal/compute/IgniteComputeImplTest.java]: done

File: modules/compute/src/test/java/org/apache/ignite/internal/compute/IgniteComputeImplTest.java [hunk 4]
```diff
@@ -243,19 +268,24 @@ private void respondWhenAskForPrimaryReplica() {
     }
 
     private void respondWhenExecutingSimpleJobLocally(ExecutionOptions executionOptions) {
-        when(computeComponent.executeLocally(eq(executionOptions), eq(testDeploymentUnits), eq(JOB_CLASS_NAME), eq("a")))
+        when(computeComponent.executeLocally(eq(executionOptions), eq(testDeploymentUnits), eq(JOB_CLASS_NAME), eq(null), eq("a")))
```
[the remark this answers is not in the document]
Comment (zstan) [reply modules/compute/src/test/java/org/apache/ignite/internal/compute/IgniteComputeImplTest.java]: done

File: modules/compute/src/test/java/org/apache/ignite/internal/compute/IgniteComputeImplTest.java [hunk 5]
```diff
@@ -243,19 +268,24 @@ private void respondWhenAskForPrimaryReplica() {
...
                 .thenReturn(completedExecution("jobResponse"));
     }
 
+    private void respondWhenExecutingSimpleJobLocally(ExecutionOptions executionOptions, CancellationToken cancellationToken) {
+        when(computeComponent.executeLocally(eq(executionOptions), eq(testDeploymentUnits), eq(JOB_CLASS_NAME),
+                eq(cancellationToken), eq("a"))).thenReturn(completedExecution("jobResponse"));
```
[the remark this answers is not in the document]
Comment (zstan) [reply modules/compute/src/test/java/org/apache/ignite/internal/compute/IgniteComputeImplTest.java]: done

File: modules/compute/src/test/java/org/apache/ignite/internal/compute/IgniteComputeImplTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (zstan) [reply modules/compute/src/test/java/org/apache/ignite/internal/compute/IgniteComputeImplTest.java]: done

File: modules/compute/src/main/java/org/apache/ignite/internal/compute/RemoteExecutionContext.java [hunk 6]
```diff
@@ -41,7 +41,12 @@ class RemoteExecutionContext<T, R> {
 
     private final AtomicReference<FailSafeJobExecution<R>> jobExecution;
 
-    RemoteExecutionContext(List<DeploymentUnit> units, String jobClassName, ExecutionOptions executionOptions, T arg) {
+    RemoteExecutionContext(
+            List<DeploymentUnit> units,
+            String jobClassName,
+            ExecutionOptions executionOptions,
+            T arg
+    ) {
```
[the remark this answers is not in the document]
Comment (zstan) [reply modules/compute/src/main/java/org/apache/ignite/internal/compute/RemoteExecutionContext.java]: done

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/client/ItAbstractThinClientTest.java [hunk 7]
```diff
@@ -229,4 +238,29 @@ public String getVal() {
             return val;
         }
     }
+
+    static class InfiniteMapReduceTask implements MapReduceTask<Void, Void, Void, Void> {
```
[the remark this answers is not in the document]
Comment (zstan) [reply modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/client/ItAbstractThinClientTest.java]: done

**Review by korlov42 — CHANGES_REQUESTED** [review]
In JIRA, Definition of Done is defined as  

> Need to add another group of execute<*> methods to IgniteCompute interface accepting optional CancellationToken as input parameter.

This PR doesn't follow suggested approach.

File: modules/api/src/main/java/org/apache/ignite/compute/IgniteCompute.java [hunk 8]
```diff
@@ -48,9 +49,29 @@ public interface IgniteCompute {
...
+     * @param cancellationToken Cancellation token or {@code null}.
+     * @param arg Argument of the job.
+     * @return Job execution object.
+     */
     <T, R> JobExecution<R> submit(
             JobTarget target,
             JobDescriptor<T, R> descriptor,
+            @Nullable CancellationToken cancellationToken,
             @Nullable T arg
     );
```
Comment (korlov42) [inline modules/api/src/main/java/org/apache/ignite/compute/IgniteCompute.java:76]: all `submit*` methods should be left intact, since they return `JobExecution` which is handle itself containing `cancel` method

File: modules/api/src/main/java/org/apache/ignite/compute/JobExecutionOptions.java [hunk 9]
```diff
@@ -36,6 +36,7 @@ public class JobExecutionOptions {
      *
      * @param priority Job execution priority.
      * @param maxRetries Number of times to retry job execution in case of failure, 0 to not retry.
+     *
```
Comment (korlov42) [inline modules/api/src/main/java/org/apache/ignite/compute/JobExecutionOptions.java:39]: unrelated change

Comment (zstan) [reply modules/api/src/main/java/org/apache/ignite/compute/JobExecutionOptions.java]: done

File: modules/api/src/main/java/org/apache/ignite/compute/IgniteCompute.java (hunk 8 above)
Comment (zstan) [reply modules/api/src/main/java/org/apache/ignite/compute/IgniteCompute.java]: done
